src/System.Net.Security/src/System/Net/_SslSessionsCache.cs

```diff
@@ -0,0 +1,201 @@
+
+            if (cached == null || cached.IsClosed || cached._Target.IsInvalid)
+            {
+                GlobalLog.Print("TryCachedCredential() Not Found, Current Cache Count = " + s_CachedCreds.Count);
+                return null;
+            }
+
+            GlobalLog.Print("TryCachedCredential() Found a cached Handle = " + cached._Target.ToString());
+
+            return cached._Target;
+        }
+
+        //
+        // The app is calling this method after starting an SSL handshake.
+        //
+        // ATTN: The thumbPrint must be from inspected and possibly cloned user Cert object or we get a security hole in SslCredKey ctor.
+        //
+        internal static void CacheCredential(SafeFreeCredentials creds, byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
+        {
+            GlobalLog.Assert(creds != null, "CacheCredential|creds == null");
+            if (creds.IsInvalid)
+            {
+                GlobalLog.Print("CacheCredential() Refused to cache an Invalid Handle = " + creds.ToString() + ", Current Cache Count = " + s_CachedCreds.Count);
+                return;
+            }
+
+            object key = new SslCredKey(thumbPrint, allowedProtocols, encryptionPolicy);
+
+            SafeCredentialReference cached = s_CachedCreds[key] as SafeCredentialReference;
+
+            if (cached == null || cached.IsClosed || cached._Target.IsInvalid)
+            {
+                lock (s_CachedCreds)
+                {
+                    cached = s_CachedCreds[key] as SafeCredentialReference;
+
+                    if (cached == null || cached.IsClosed)
+                    {
+                        cached = SafeCredentialReference.CreateReference(creds);
+
+                        if (cached == null)
+                        {
+                            // Means the handle got closed in between, return it back and let caller deal with the issue.
+                            return;
+                        }
+
+                        s_CachedCreds[key] = cached;
+                        GlobalLog.Print("CacheCredential() Caching New Handle = " + creds.ToString() + ", Current Cache Count = " + s_CachedCreds.Count);
+
+                        //
+                        // A simplest way of preventing infinite cache grows.
+                        //
+                        // Security relief (DoS):
+                        //     A number of active creds is never greater than a number of _outstanding_
+                        //     security sessions, i.e. SSL connections.
+                        //     So we will try to shrink cache to the number of active creds once in a while.
+                        //
+                        //    We won't shrink cache in the case when NO new handles are coming to it.
+                        //
+                        if ((s_CachedCreds.Count % c_CheckExpiredModulo) == 0)
```

Comment:
Why mod instead of `>=`?  is it not possible to have two add at close enough to the same time to skip right over the modulo?